### PR TITLE
feature: add range functionality to `fix_time`

### LINKED ISSUE #27: Add range functionality to `fix_time`
See [here](https://github.com/ANNUBS/annubes/blob/34d9787ba81ac0a4e949aab539444b7e887857a1/annubes/task.py#L25) for the original implementation in annubes.

### PR DESCRIPTION
**Pull request by gcroci2**
The functionality was already there for all the times defined in the `self.timing` attribute, so for `fix_time` as well. I've added the doc string and the tests.

There is an issue for the "until" functionality, but it's not trivial so I've opened a separate issue for it (#64).

### CODE REVIEW

File: neurogym/core.py [hunk 1]
```diff
@@ -254,6 +255,10 @@ def sample_time(self, period):
             elif dist == "until":
                 # set period duration such that self.t_end[period] = args
                 t = args - self.tmax
+                if t < 0:
+                    msg = f"""Invalid 'until' time for period {period}. Current max time: {self.tmax},
+                    Requested end time: {args}"""
+                    raise ValueError(msg)
```
Comment (DaniBodor) [inline neurogym/core.py]: this avoids printing the newline and indents in the error message.

```suggestion
                    msg = (
                        f"Invalid 'until' time for period {period}. Current max time: {self.tmax},",
                        f"Requested end time: {args}",
                    )
                    raise ValueError(msg)
```

File: neurogym/core.py [hunk 2]
```diff
@@ -305,6 +310,9 @@ def add_period(
 
         if duration is None:
             duration = self.sample_time(period)
+            self._duration[period] = duration
+        else:
+            self._duration[period] = duration
```
Comment (DaniBodor) [inline neurogym/core.py]:
```suggestion
        self._duration[period] = duration
```

File: neurogym/envs/annubes.py [hunk 3]
```diff
@@ -24,7 +26,18 @@ class AnnubesEnv(TrialEnv):
...
+            - A callable: Function that returns the duration when called.
+            - A list of numbers: Random choice from the list.
+            - A tuple specifying a distribution:
+                - ("uniform", min, max): Uniform distribution between min and max.
+                - ("choice", [options]): Random choice from the given options.
+                - ("truncated_exponential", [parameters]): Truncated exponential distribution.
+                - ("constant", value): Always returns the given value.
+                - ("until", end_time): Sets duration to reach the specified end time.
+            The final duration is rounded down to the nearest multiple of the simulation timestep (dt).
+            Note that the duration of each input and output signal is increased by this time.
```
Comment (DaniBodor) [inline neurogym/envs/annubes.py:40]: I have the impression that there is a bit too much going on in this one parameter. There are too many options to choose from.

Maybe add a mixin class for the fixation settings, something like this:

NB: it is unclear to me what the "until" and "callable" modes are.

```python
@dataclass
class FixationSettings:
    """Mix-in class for AnnubesEnv that specifies what the fixation period should look like.

    Args:
        intensity: Intensity during the fixation period.
        mode: Options are: "constant", "uniform", "choice", "truncated_exponential", "until".
        values: values to determine the length of the fixation period in ms.
            for "constant" or "until" modes, a single value is expected.
            for "uniform" mode, the min and max values of the uniform distribution to randomly choose from are expected.
            for "choice", a list of options to randomly choose from is expected.
            for "truncated_exponential" modes, a list of parameters for the distribution is expected.
            for "callable", a function is expected.
    """

    intensity: int | float = 0
    mode: Literal["constant", "uniform", "choice", "truncated_exponential", "until", "callable"] = "constant"
    values: float | tuple[float] | list[float] | Callable = 500

    def __post_init__(self):
        msg = ""
        single_value_expected = ["constant", "until"]
        if self.mode in single_value_expected:
            if not isinstance(self.values, float):
                msg = f"The {self.mode} mode expects a single value, but {self.values} ({type(self.mode)}) was given."

        elif self.mode == "callable":
            if not isinstance(self.values, Callable):
                msg = f"The {self.mode} mode expects a function, but {self.values} ({type(self.mode)}) was given."

        elif not isinstance(self.values, tuple | list):
            msg = f"The {self.mode} mode expects a list of values, but {self.values} ({type(self.mode)}) was given."

        if msg:
            raise ValueError(msg)
```

**Review by DaniBodor — CHANGES_REQUESTED** [review]
Thanks for these options for the duration of the fixation time.
As commented below, I find the current implementation a bit confusing. I made a suggestion to offload those settings into a mixin class. What do you think of that idea?

File: neurogym/envs/annubes.py (hunk 3 above)
Comment (gcroci2) [reply neurogym/envs/annubes.py:40]: Such functionalities were already implemented within the core module (see [here](https://github.com/ANNUBS/neurogym/blob/29368d5f319e544b141b93867cd4227c43a4af29/neurogym/core.py#L236)), so I just added a docstring about them. I think that in the future we may want to improve this directly (included "until" and "callable" modes) within the core module, but for now I'd use the functionality as it is. 

As for the mixin class, I'd prefer a tuple for specifying a distribution instead two different args (`mode` and `values`). I do agree that it quite a lot of text now,  but I think that this is related to the issue about having too many input parameters for the class, and thus I think is better to solve it all in a go in a later stage. I opened #68 for this.

Comment (DaniBodor) [reply neurogym/envs/annubes.py:40]: I understand. At this point, the "until" and "callable" functionality are still unclear. Maybe add a few words explaining these?

**Review by DaniBodor — APPROVED** [review]
I'm approving this already, but please see my comment on the docstring.

File: neurogym/envs/annubes.py (hunk 3 above)
Comment (gcroci2) [reply neurogym/envs/annubes.py:40]: I opened #69

**Pull request merged**